syntax = "proto3";

service InventoryService {
  rpc CheckAvailability (InventoryRequest) returns (InventoryResponse);
  rpc ReserveItems (ReserveRequest) returns (ReserveResponse);
  rpc ReleaseItems (ReleaseRequest) returns (ReleaseResponse);
}

message InventoryRequest {
  map<string, int32> items = 1;
}

message InventoryResponse {
  map<string, bool> availability = 1;
}

message ReserveRequest {
  map<string, int32> items = 1;
}

message ReserveResponse {
  bool overallSuccess = 1;
  map<string, ReserveStatus> results = 2;
}

message ReserveStatus {
  bool success = 1;
  string message = 2;
}

message ReleaseRequest {
  map<string, int32> items = 1;
}

message ReleaseResponse {
  bool overallSuccess = 1;
  map<string, string> messages = 2;
}
